text/0000-module-unification.md

```diff
@@ -0,0 +1,837 @@
+- Start Date: 2016-05-09
+- RFC PR:
+- Ember Issue:
+
+> Note: This RFC replaces the closely related RFC for [Module
+Normalization](https://github.com/emberjs/rfcs/pull/124). As discussed in the
+[Alternatives](#alternatives) section below, many concepts are shared between
+the two proposals, but there is also a fundamental difference.
+
+# Summary
+
+Create a unified pattern for organizing and naming modules in Ember projects
+that is deterministic, extensible, and ergonomic.
+
+# Motivation
+
+Ember CLI's conventions for project layouts and file naming are central to every
+Ember developer's experience. It's crucial to get both the technical and
+ergonomic details right.
+
+The existing conventions used by Ember CLI have evolved gradually and
+organically over the years. Ember CLI and its predecessor Ember App Kit were
+early adopters of ES modules and have always leveraged strong conventions to
+deduce an understanding of modules based on their locations. Ember CLI's
+resolver encodes those conventions to enable run-time module resolutions.
+
+The current system works fairly well, but has some complexities and
+inconsistencies that both steepen its learning curve and limit its technical
+potential.
+
+Drawbacks include:
+
+* Confusion over which of two orthogonal approaches to use for organizing
+  modules:
+
+  * classic - modules are organized at the top-level by "type"
+    (`components`, `templates`, etc.) and then by namespace and name.
+
+  * pods - modules are organized by namespace, then name, then type.
+
+* Addons define modules to be merged into an application through a special `app`
+  directory. These public modules are typically private modules that are
+  imported and re-exported, which introduces an extra module per export and an
+  extra level of abstraction to learn.
+
+* Because addons' modules are mixed into an application, there's the possibility
+  of naming collisions between two addons or an addon and its consuming
+  application.
+
+* Modules don't have a clear sense of "locality", which prevents the ability to
+  declare modules that are available only in a "local" namespace (this as-yet
+  unsupported feature has been called "local lookup").
+
+* Resolution rules that are declared only in JavaScript are difficult to
+  analyze and optimize.
+
+* Module resolution is inefficient due to the number of potential places to
+  lookup a particular module by name.
+
+Recognizing these drawbacks, the Core Team compiled a set of
+[design constraints](https://github.com/emberjs/core-notes/blob/master/ember.js/2016-01/january-22.md#summary-of-pods-design-constraints)
+for a rethink of Ember's approach to modules:
+
+1. Reasonable branching factor. Users should see a reasonable number of items at any given level in their hierarchy. Flattening out too much results in an unreasonably large number of items.
```

Comment:
@rtm No doubt there will always be terminology that boils down to a judgment call, even in matters of life or death—in the U.S. justice system, you must find someone guilty "beyond reasonable doubt," for example.

In this case, I think of the file system rules a bit like a hashing algorithm. Which algorithm is best has a lot to do with the shape and distribution of the data being hashed. You want a nice distribution while minimizing pathological cases.

Given that there are so many apps in the wild, this is exactly the point of the RFC process, and why Rob has worked so hard on a migrator tool. Our hope is that people will use the migrator with the proposed rules, examine the output, and share any pathological cases that they feel might be a cause to revisit certain rules—after, of course, [giving it five minutes](https://signalvnoise.com/posts/3124-give-it-five-minutes) and not just disliking it because it's different.